File: packages/core/src/backtest/metrics.ts
import { mean, median, quantile, stddev, sum } from '../util/stats.js';

export interface TradeResult {
  signalId: string;
  tokenMint: string;
  entryAt: number;
  exitAt: number;
  sizeUsd: number;
  /** Net of fees, slippage and failure costs. */
  pnlUsd: number;
  pnlPct: number;
  quotedPriceUsd: number;
  filledPriceUsd: number;
  rugged: boolean;
  /** Did the token expand ≥2x within the horizon after the signal (whether or not captured)? */
  majorMove: boolean;
}

export interface ProfitabilityMetrics {
  trades: number;
  netExpectedValuePct: number;
  netPnlUsd: number;
  profitFactor: number;
  winRate: number;
  maxDrawdownPct: number;
  sharpe: number;
  sortino: number;
  calmar: number;
  medianTradePct: number;
  tailLossPct: number; // 5th percentile trade return
  opportunityCapture: number;
  falsePositiveRate: number;
  executionSlippagePct: number;
  /** 95% CI on mean trade return via bootstrap. */
  evConfidenceInterval: [number, number];
}

export function equityCurve(trades: TradeResult[], startEquity: number): number[] {
  const curve = [startEquity];
  let eq = startEquity;
  for (const t of [...trades].sort((a, b) => a.exitAt - b.exitAt)) {
    eq += t.pnlUsd;
    curve.push(eq);
  }
  return curve;
}

export function maxDrawdown(curve: number[]): number {
  let peak = curve[0] ?? 0;
  let dd = 0;
  for (const v of curve) {
    if (v > peak) peak = v;
    if (peak > 0) dd = Math.max(dd, (peak - v) / peak);
  }
  return dd * 100;
}

export function bootstrapCI(xs: number[], iters = 500, seed = 7): [number, number] {
  if (xs.length === 0) return [0, 0];
  let a = seed >>> 0;
  const rnd = () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
  const means: number[] = [];
  for (let i = 0; i < iters; i++) {
    let s = 0;
    for (let j = 0; j < xs.length; j++) s += xs[Math.floor(rnd() * xs.length)];
    means.push(s / xs.length);
  }
  return [quantile(means, 0.025), quantile(means, 0.975)];
}

/** Profitability metrics (spec §20). `missedMoves` = major moves the system never alerted on. */
export function computeMetrics(trades: TradeResult[], startEquity: number, missedMoves = 0, periodsPerYear = 365): ProfitabilityMetrics {
  const rets = trades.map((t) => t.pnlPct);
  const wins = trades.filter((t) => t.pnlUsd > 0);
  const losses = trades.filter((t) => t.pnlUsd < 0);
  const gp = sum(wins.map((t) => t.pnlUsd));
  const gl = -sum(losses.map((t) => t.pnlUsd));
  const curve = equityCurve(trades, startEquity);
  const dd = maxDrawdown(curve);
  const dailyRet = rets.map((r) => r / 100);
  const sd = stddev(dailyRet);
  const downside = Math.sqrt(mean(dailyRet.map((r) => Math.min(0, r) ** 2))) || 0;
  const annual = mean(dailyRet) * periodsPerYear;
  const total = curve.length > 1 ? (curve[curve.length - 1] - curve[0]) / curve[0] : 0;
  const captured = trades.filter((t) => t.majorMove).length;
  return {
    trades: trades.length,
    netExpectedValuePct: mean(rets),
    netPnlUsd: sum(trades.map((t) => t.pnlUsd)),
    profitFactor: gl === 0 ? (gp > 0 ? Infinity : 0) : gp / gl,
    winRate: trades.length ? wins.length / trades.length : 0,
    maxDrawdownPct: dd,
    sharpe: sd > 0 ? (mean(dailyRet) / sd) * Math.sqrt(periodsPerYear) : 0,
    sortino: downside > 0 ? (mean(dailyRet) / downside) * Math.sqrt(periodsPerYear) : 0,
    calmar: dd > 0 ? (total * 100) / dd : 0,
    medianTradePct: median(rets),
    tailLossPct: quantile(rets, 0.05),
    opportunityCapture: captured + missedMoves > 0 ? captured / (captured + missedMoves) : 0,
    falsePositiveRate: trades.length ? trades.filter((t) => t.pnlPct <= -10 || t.rugged).length / trades.length : 0,
    executionSlippagePct: mean(trades.map((t) => (t.quotedPriceUsd > 0 ? Math.abs(t.filledPriceUsd / t.quotedPriceUsd - 1) * 100 : 0))),
    evConfidenceInterval: bootstrapCI(rets),
  };
}

/** Signal half-life: how quickly forward edge decays after detection, from (delayMs, meanReturn) samples. */
export function signalHalfLife(samples: Array<{ delayMs: number; meanReturnPct: number }>): number | null {
  const sorted = [...samples].sort((a, b) => a.delayMs - b.delayMs);
  if (sorted.length < 2 || sorted[0].meanReturnPct <= 0) return null;
  const half = sorted[0].meanReturnPct / 2;
  for (const s of sorted) if (s.meanReturnPct <= half) return s.delayMs;
  return null;
}

File: research/meme_research/metrics.py
"""Profitability metrics (spec §20). Input: a DataFrame of closed trades with columns
`pnl_pct`, `pnl_usd`, `size_usd`, `rugged`, `major_move`, `quoted_price`, `filled_price`, `exit_at`."""
from __future__ import annotations

import numpy as np
import pandas as pd


def _bootstrap_ci(x: np.ndarray, iters: int = 1000, seed: int = 7) -> tuple[float, float]:
    if len(x) == 0:
        return (0.0, 0.0)
    rng = np.random.default_rng(seed)
    means = rng.choice(x, size=(iters, len(x)), replace=True).mean(axis=1)
    return (float(np.quantile(means, 0.025)), float(np.quantile(means, 0.975)))


def max_drawdown_pct(equity: np.ndarray) -> float:
    if len(equity) == 0:
        return 0.0
    peak = np.maximum.accumulate(equity)
    with np.errstate(divide="ignore", invalid="ignore"):
        dd = np.where(peak > 0, (peak - equity) / peak, 0.0)
    return float(dd.max() * 100)


def profitability_metrics(trades: pd.DataFrame, start_equity: float, missed_moves: int = 0, periods_per_year: int = 365) -> dict:
    if trades.empty:
        return {"trades": 0, "net_ev_pct": 0.0, "profit_factor": 0.0, "win_rate": 0.0, "max_drawdown_pct": 0.0, "sharpe": 0.0, "sortino": 0.0, "calmar": 0.0, "median_trade_pct": 0.0, "tail_loss_pct": 0.0, "opportunity_capture": 0.0, "false_positive_rate": 0.0, "execution_slippage_pct": 0.0, "ev_ci": (0.0, 0.0)}
    t = trades.sort_values("exit_at") if "exit_at" in trades else trades
    rets = t["pnl_pct"].to_numpy(dtype=float)
    pnl = t["pnl_usd"].to_numpy(dtype=float)
    gp = pnl[pnl > 0].sum()
    gl = -pnl[pnl < 0].sum()
    equity = start_equity + np.cumsum(pnl)
    equity = np.concatenate([[start_equity], equity])
    dd = max_drawdown_pct(equity)
    r = rets / 100
    sd = r.std(ddof=1) if len(r) > 1 else 0.0
    downside = np.sqrt(np.mean(np.minimum(r, 0) ** 2))
    total = (equity[-1] - equity[0]) / equity[0]
    captured = int(t["major_move"].sum()) if "major_move" in t else 0
    slip = np.abs(t["filled_price"] / t["quoted_price"] - 1) * 100 if {"filled_price", "quoted_price"} <= set(t.columns) else pd.Series([0.0])
    rugged = t["rugged"] if "rugged" in t else pd.Series(False, index=t.index)
    return {
        "trades": int(len(t)),
        "net_ev_pct": float(rets.mean()),
        "net_pnl_usd": float(pnl.sum()),
        "profit_factor": float(gp / gl) if gl > 0 else float("inf") if gp > 0 else 0.0,
        "win_rate": float((pnl > 0).mean()),
        "max_drawdown_pct": dd,
        "sharpe": float(r.mean() / sd * np.sqrt(periods_per_year)) if sd > 0 else 0.0,
        "sortino": float(r.mean() / downside * np.sqrt(periods_per_year)) if downside > 0 else 0.0,
        "calmar": float(total * 100 / dd) if dd > 0 else 0.0,
        "median_trade_pct": float(np.median(rets)),
        "tail_loss_pct": float(np.quantile(rets, 0.05)),
        "opportunity_capture": captured / (captured + missed_moves) if captured + missed_moves > 0 else 0.0,
        "false_positive_rate": float(((rets <= -10) | rugged.to_numpy()).mean()),
        "execution_slippage_pct": float(np.mean(slip)),
        "ev_ci": _bootstrap_ci(rets),
    }


def signal_half_life(samples: pd.DataFrame) -> float | None:
    """`samples` has `delay_ms` and `mean_return_pct`; returns delay at which edge halves."""
    s = samples.sort_values("delay_ms")
    if s.empty or s.iloc[0]["mean_return_pct"] <= 0:
        return None
    half = s.iloc[0]["mean_return_pct"] / 2
    hit = s[s["mean_return_pct"] <= half]
    return float(hit.iloc[0]["delay_ms"]) if not hit.empty else None
